include/circt/Dialect/OM/OMOps.td

```diff
@@ -195,4 +197,104 @@ def ListCreateOp : OMOp<"list_create", [Pure, SameTypeOperands]> {
 }
 
+def TupleCreateOp : OMOp<"tuple_create", [Pure, InferTypeOpInterface]> {
+  let summary = "Create a tuple of values";
+  let description = [{
+    Create a tuple from a sequence of inputs.
+
+    ```
+    %tuple = om.tuple_create %a, %b, %c : !om.ref, !om.string, !om.list<i32>
+    ```
+  }];
+
+  let arguments = (ins Variadic<AnyType>:$inputs);
+  let results = (outs
+    TupleOf<[AnyType]>:$result
+  );
+
+  let assemblyFormat = [{
+    $inputs `:` type($inputs) attr-dict
+  }];
+
+  let extraClassDeclaration = [{
+    // Implement InferTypeOpInterface.
+    static ::mlir::LogicalResult inferReturnTypes(
+      ::mlir::MLIRContext *context, ::std::optional<::mlir::Location> location,
+      ::mlir::ValueRange operands, ::mlir::DictionaryAttr attributes,
+      ::mlir::OpaqueProperties,
+      ::mlir::RegionRange regions,
+      ::llvm::SmallVectorImpl<::mlir::Type> &inferredReturnTypes) {
+      ::llvm::SmallVector<::mlir::Type> types;
+      for (auto op: operands) 
+        types.push_back(op.getType());
+      inferredReturnTypes.push_back(mlir::TupleType::get(context, types));
+      return ::mlir::success();
+    }
+   }];
+
+}
+
+def TupleGetOp : OMOp<"tuple_get", [Pure, InferTypeOpInterface]> {
+  let summary = "Extract a value from a tuple";
+  let description = [{
+    Extract a value from a tuple.
+
+    ```
+    %value = om.tuple_get %a[0] : tuple<!om.ref, !om.string, !om.list<i32>>
+    ```
+  }];
+
+  let arguments = (ins
+    TupleOf<[AnyType]>:$tuple,
+    I32Attr:$index
+  );
+
+  let results = (outs
+    AnyType:$result
+  );
+
+  let assemblyFormat = [{
+    $tuple `[` $index `]` `:` type($tuple) attr-dict
+  }];
+
+  let extraClassDeclaration = [{
+    // Implement InferTypeOpInterface.
+    static ::mlir::LogicalResult inferReturnTypes(
+      ::mlir::MLIRContext * context, ::std::optional<::mlir::Location> location,
+      ::mlir::ValueRange operands, ::mlir::DictionaryAttr attributes,
+      ::mlir::OpaqueProperties, ::mlir::RegionRange regions,
+      ::llvm::SmallVectorImpl<::mlir::Type> & inferredReturnTypes) {
+    auto idx = attributes.getAs<::mlir::IntegerAttr>("index");
+    inferredReturnTypes.push_back(
+        operands[0]
+            .getType()
+            .cast<mlir::TupleType>()
+            .getTypes()[idx.getValue().getLimitedValue()]);
+    return ::mlir::success();
+  }
+   }];
+}
+
+def StringKeyPair: Type<CPred<"::circt::om::isMapKeyValuePairType($_self)">,
+                        "a pair whose first element is a string",
+                        "::mlir::TupleType">;
+
+def MapCreateOp : OMOp<"map_create", [Pure, SameTypeOperands]> {
+  let summary = "Create a map";
+  let description = [{
+    Creates a map from a sequence of inputs.
+
+    ```
+    %map = om.map_create %e1, %e2 : i8
+    ```
+    where `%map` has `!om.map<i8>` type.
+  }];
+
+  let arguments = (ins Variadic<StringKeyPair>:$inputs);
+  let results = (outs
+    MapType:$result
+  );
+  let hasCustomAssemblyFormat = 1;
```

Comment:
Have you tried using the generated parser/printer now that `$inputs` is just a plain old list of operands?

Reply:
Thanks, i tried but it seems it doesn't work since `type($inputs)` requires enumerating all operand types even when `SameOperandTypes` trait.